### PR TITLE
Server vulnerability assessment resource

### PR DESCRIPTION
**Pull request by martenbohlin**
Adds a new resource azurerm_server_vulnerability_assessment that installes the vulnerability assessment on a VM

### CODE REVIEW

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:36]: Seems ServerVulnerabilityAssessment supports servers, which can be of type “Microsoft.Compute/virtualMachines”, But also “[Microsoft.HybridCompute/machines](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2020-08-02/HybridCompute.json)”.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:39]: Seems ServerVulnerabilityAssessment can be updated from server A to server B. So ForceNew isn't required. Could you double check whether it could be updated in-place?

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:36]: I can not find any information on how to create or work with Microsoft.HybridCompute resources from terraform. @neil-yechenwei do you have any information on how to work with HybridCompute in terraform? 
I will need some help in order to know how to setup something where I could test and verify HybridCompute support.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:39]: I will take a look at this.

File: website/docs/r/azurerm_server_vulnerability_assessment.markdown
Comment (sharebear) [inline website/docs/r/azurerm_server_vulnerability_assessment.markdown:47]: The documentation for `azurerm_virtual_machine` states

> The azurerm_virtual_machine resource has been superseded by the azurerm_linux_virtual_machine and azurerm_windows_virtual_machine resources.

Should examples in documentation be updated to one of these?

**Comment by sharebear** [comment]
Fun, I was just looking at the API for this yesterday to see how much work it would be to implement :)

This is getting towards a functional question rather than a PR comment, but have I understood correctly that you need to have activated Azure Defender for Servers enabled on the Subscription in order to use Qualys;

1. Should there be a note about this in the documentation?
2. How does the resource behave if the Subscription does _not_ have Azure Defender enabled?
3. Do the acceptance tests run in a Subscription that has Azure Defender enabled?

**Comment by martenbohlin** [comment]
> This is getting towards a functional question rather than a PR comment, but have I understood correctly that you need to have activated Azure Defender for Servers enabled on the Subscription in order to use Qualys;
> 
> 1. Should there be a note about this in the documentation?
> 2. How does the resource behave if the Subscription does _not_ have Azure Defender enabled?
> 3. Do the acceptance tests run in a Subscription that has Azure Defender enabled?

Good  questions. It would be interesting to know what would happen if Azure Defender is disabled. If I could get help to test that it would be great since the company I work for require defender to be enabled on all subscriptions. (Should be enough to run the test TestAccServerVulnerabilityAssessment_basic)

File: website/docs/r/azurerm_server_vulnerability_assessment.markdown
Comment (martenbohlin) [reply website/docs/r/azurerm_server_vulnerability_assessment.markdown]: Thanks, I have updated the documentation

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:39]: @neil-yechenwei I can not see how I would do a in-place update with the API.The API has the method **CreateOrUpdate**, but that method takes the parameters resourceGroupName, resourceNamespace, resourceType, and resourceName to identify the **server** to create or update the vulnerability assessment on. So I do not think it is possible to specify a specific vulnerability assessment and move that. Instead you need to delete the old and install a new.
Do you agree, or am I missing something?

Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:39]: Yes. I assume you're right.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:36]: Terraform doesn't support Microsoft.HybridCompute/machines for now. I assume you can create/test this resource with [rest api](https://docs.microsoft.com/en-us/rest/api/hybridcompute/machines/createorupdate). I assume the resource `azurerm_server_vulnerability_assessment` could be made more generic to support both of them if ServerVulnerabilityAssessment API also supports Microsoft.HybridCompute/machines?

Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:36]: I have added support for Microsoft.HybridCompute/machines (Azure ARC). But I have kept the name virtual_machine_id for the property since that seems to be most consistent with other resources. Or does anybody have any suggestions on a better name?

**Comment by martenbohlin** [comment]
> This is getting towards a functional question rather than a PR comment, but have I understood correctly that you need to have activated Azure Defender for Servers enabled on the Subscription in order to use Qualys;
> 
> 1. Should there be a note about this in the documentation?
> 2. How does the resource behave if the Subscription does _not_ have Azure Defender enabled?
> 3. Do the acceptance tests run in a Subscription that has Azure Defender enabled?

@sharebear 
I was able to test this my self. So here are the answers:
1. I added a note in the documentation that Azure Defender must be enabled.
2. You get the following error message back from the API that is show:`Subscription is not in the Standard or Standard Trial subscription plan. Please upgrade to use this feature.`
3. I am guessing here, but there is tests for the resource azurerm_mssql_server_vulnerability_assessment so I belive Azure Defender is enabled where the integration tests run.

**Comment by sharebear** [comment]
> @sharebear
> I was able to test this my self. So here are the answers:
> 
> 1. I added a note in the documentation that Azure Defender must be enabled.
> 2. You get the following error message back from the API that is show:`Subscription is not in the Standard or Standard Trial subscription plan. Please upgrade to use this feature.`
> 3. I am guessing here, but there is tests for the resource azurerm_mssql_server_vulnerability_assessment so I belive Azure Defender is enabled where the integration tests run.

Awesome, I think that's the best possible answer for these questions.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (sharebear) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:155]: Typo here `parseVirtualMachineId` however.... There already exists a parser for virtualMachines `azurerm/internal/services/compute/parse/virtual_machine.go:41` you should reuse that. This however does not support Arc.

If you want to support Arc then I would look into using the generator for resource id parsers/validators `azurerm/internal/tools/generator-resource-id/README.md`. Then I would split the input parameters into conflicting `virtual_machine_id` and `hybrid_machine_id` allowing you to use the generated validators for each id type. Although _personally_ I would probably drop the Arc support from this PR to keep it small, replacing this parser with the existing parser. Then if Arc support is desired, implement that as a new PR on top of these changes.

Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:155]: @neil-yechenwei is Arc support something you need, or can we wait with that?

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (sharebear) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:145]: I would file this as a bug against the upstream SDK. If the API gives a 202 on "success" then it should probably be added to the list of accepted codes in DeleteResponder.

Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:146]: [Bug added to the SDK](https://github.com/Azure/azure-sdk-for-go/issues/14120)

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go]: @sharebear I would appreciate if you took a look at my changes in commit 2ffd690 and let me know if you think they look good.

Comment (sharebear) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go]: It's probably best to wait and see what the maintainers say, and I'm still very new to golang and writing providers. That said, I have a gut feeling that the implementation feels a little complicated compared to other resources I've looked at.... what if read looked like the following

```
targetId = ""
virtualMachineId, err := computeParse.VirtualMachineID(d.Id())
if virtualMachineId != nil {
	targetId = virtualMachineId.ID()
	d.Set("virtual_machine_id", targetId)
}

if targetId == "" {
	hybridMachineId, err := computeParse.HybridMachineID(d.Id())
	if hybridMachineId != nil {
		targetId = hybridMachineId.ID()
		d.Set("hybrid_machine_id", targetId)
	}
}

// TODO: Return err if targetId still emtpy

// do client.Get
```

I think create would be even easier to simplify, then you could drop the whole TargetId struct.... disclaimer, I'm shooting from the hip here, no idea if this would work in practice.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:43]: How about `target_machine_id`?

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:49]: Is it possible to merge "virtual_machine_id" and "hybrid_machine_id" to one property "target_machine_id"?

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:177]: Does it make sense to parse it by judging whether the "target_machine_id" exists the resource id segment "Microsoft.HybridCompute/machines" after merged these two properties to one property "target_machine_id"?

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:19]: I assume here should be "example-resources".

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:24]: I assume here should be "example-vnet".

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:31]: I assume here should be "example-subnet".

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:38]: I assume here should be "example-nic".

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:50]: I assume here should be "example-vm".

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:111]: Suggest to also add an example of Microsoft.HybridMachine.

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:107]: I assume it should be ”Server Vulnerability Assessments“

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:20]: I assume here should be "West Europe"

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:91]: I assume here should be "acctestVM-%[1]d"

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:79]: I assume here should be "acctestNIC-%[1]d"

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:72]: I assume here should be "acctestSubnet-%[1]d"

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:65]: I assume here should be "acctestVNet-%[1]d"

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:60]: I assume here should be "acctestRG-sva-%[1]d"

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:53]: I assume the function name should be "template" not "basicCfg"

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:36]: Seems this function also should be compatible with Hybrid Machine, right?

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [inline azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:18]: I assume it's necessary to add a test case for requiresImport, right?

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:6]: It could be updated to "Manages an Azure Vulnerability Assessment (Qualys) to a VM or Azure ARC server" to align the style with other existing resources. Does it make sense?

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:11]: It could be updated to "Manages an Azure Server Vulnerability Assessment (Qualys) to a VM." to align the style with other existing resources. Does it make sense?

**Comment by neil-yechenwei** [comment]
@martenbohlin , Thanks for the update. I have left a couple of comments, please have a look, thanks

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go]: Agree, my problem is that the API needs a couple specific parts of the id and not the entire id. Will keep like this and see what the maintainers say as you suggested.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:49]: Yes I had it like that fist. But then got the suggestion to split into separate properties. Do you have any strong arguments to just have one property?
I like the name virtual_machine_id since that is the most commonly used when you reference a VM.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:177]: Yes, depending on if we merge the properties or not.

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (martenbohlin) [reply website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:111]: I do not think it is necessary since it is just to copy paste whatever ID your resource has.

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (martenbohlin) [reply website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:20]: Any reason you think we should use Europe instead of US?

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:36]: I have no way to automatically create a Hybrid Machine. I have only been able to verify that manually.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:18]: On line 32 I have data.ImportStep() that will verify that import works as part of the TestAccServerVulnerabilityAssessment_basic test case.

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [reply website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:111]: I saw an example that it lists all possible samples. It's fine to keep it for now.

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (neil-yechenwei) [reply website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:20]: I assume it should be the the code style of terraform. You can reference other [resources](https://raw.githubusercontent.com/terraform-providers/terraform-provider-azurerm/master/website/docs/r/stack_hci_cluster.html.markdown) as sample.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:36]: Maybe you can manually create it first via [azure rest api](https://docs.microsoft.com/en-us/rest/api/hybridcompute/machines/createorupdate) and configure the resource id of this resource in test case. Then you can verify it in test case to complete the Exists function. Does it make sense?

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (neil-yechenwei) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:18]: I assume it's not enough. Suggest to add "requiresImport" tc like other [resources](https://github.com/terraform-providers/terraform-provider-azurerm/blob/59066bd464d5f5cce8eddc3e4d5e4f2fb0842bb4/azurerm/internal/services/azurestackhci/stack_hci_cluster_resource_test.go#L34).

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:36]: That will be a bit to complicated for me. But if you want to give it a go I think we can include it.

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (martenbohlin) [reply website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:20]: There is a wide range of locations are used in the documentation, West US is used in 123 other files. So keeping it for now.

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (martenbohlin) [reply website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:111]: That example imports on something else then the ID full id, so I think that is different since this only supports importing on the ID.

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go
Comment (martenbohlin) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource_test.go:18]: I will tak a look at this

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (katbyte) [reply website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:20]: this is fine, thou fwiw we do typically default to west europe

File: azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go
Comment (katbyte) [reply azurerm/internal/services/securitycenter/server_vulnerability_assessment_resource.go:49]: I think its better to have it split as it allows for better validation and becomes more obvious, alternatively it might make sense to split this into two resources..

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (katbyte) [reply website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:111]: I'd like us to show both please

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (katbyte) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:4]: as this is in the security center service could we rename this to
```suggestion
page_title: "Azure Resource Manager: azurerm_security_centre_server_vulnerability_assessment"
```

File: website/docs/r/azurerm_server_vulnerability_assessment.html.markdown
Comment (katbyte) [inline website/docs/r/azurerm_server_vulnerability_assessment.html.markdown:13]: could we include a link here to the docs on how to get started with this?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @martenbohlin - i've given this a review and overall while its good i've let some comments inline to address before merge

**Comment by martenbohlin** [comment]
Thanks for the feedback @katbyte ! I think all of your suggestions were great and have implemented them.

**Comment by manicminer** [comment]
@martenbohlin I've tested this locally and it looks good. Could you resolve the merge conflict and I will retest? Thanks!

**Comment by manicminer** [comment]
My bad, there wasn't a conflict after all, but thanks for merging in master.

**Review by manicminer — APPROVED** [review]
Tests are passing, this looks good to merge

Screenshot 2021-03-09 at 20 13 54

**Pull request merged**
